src/main/java/co/cask/publisher/PubSubOutputFormat.java

```diff
@@ -0,0 +1,218 @@
+      .build();
+
+    return new PubSubRecordWriter(publisher, errorThreshold);
+  }
+
+  private RetrySettings getRetrySettings(int maxRetryTimeout) {
+    // only maxRetryTimeout is configurable by user, setting defaults for other parameters
+    return RetrySettings.newBuilder()
+      .setInitialRetryDelay(Duration.ofMillis(100))
+      .setRetryDelayMultiplier(2)
+      .setMaxRetryDelay(Duration.ofSeconds(2))
+      .setTotalTimeout(Duration.ofSeconds(maxRetryTimeout))
+      .setInitialRpcTimeout(Duration.ofSeconds(1))
+      .setMaxRpcTimeout(Duration.ofSeconds(5))
+      .build();
+  }
+
+  private BatchingSettings getBatchingSettings(long countSize, long bytesThreshold, long delayThreshold) {
+    BatchingSettings.Builder batchingSettingsBuilder = BatchingSettings.newBuilder();
+    batchingSettingsBuilder.setElementCountThreshold(countSize);
+    batchingSettingsBuilder.setRequestByteThreshold(bytesThreshold);
+    batchingSettingsBuilder.setDelayThreshold(Duration.ofMillis(delayThreshold));
+    return batchingSettingsBuilder.build();
+  }
+
+  /**
+   * Writer publishes messages to PubSub using the passed Publisher,
+   * batching and retrying are performed by Publisher based on how it is configured, If publishing a message fails
+   * we maintain error count and we throw exception when this error count exceeds a threshold.
+   */
+  public class PubSubRecordWriter extends RecordWriter<NullWritable, Text> {
+    private final Publisher publisher;
+    private final AtomicLong failures;
+    private final AtomicReference<Throwable> error;
+    private final long errorThrehold;
+    private final Set<ApiFuture> futures;
+
+    public PubSubRecordWriter(Publisher publisher, long errorThreshold) throws IOException {
+      this.publisher = publisher;
+      this.error = new AtomicReference<>();
+      this.errorThrehold = errorThreshold;
+      this.failures = new AtomicLong(0);
+      this.futures = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public void write(NullWritable key, Text value) throws IOException {
+      handleErrorIfAny();
+      ByteString data = ByteString.copyFromUtf8(value.toString());
+      PubsubMessage message = PubsubMessage.newBuilder().setData(data).build();
+      ApiFuture future = publisher.publish(message);
+      futures.add(future);
+      ApiFutures.addCallback(future, new ApiFutureCallback<String>() {
+        @Override
+        public void onFailure(Throwable throwable) {
+          error.set(throwable);
+          failures.incrementAndGet();
+          futures.remove(future);
+        }
+
+        @Override
+        public void onSuccess(String s) {
+          futures.remove(future);
+        }
+      });
+    }
+
+    private void handleErrorIfAny() throws IOException {
+      if (failures.get() > errorThrehold) {
+        throw new IOException(
+          String.format("Publishing to PubSub failed for %s records due to %s", failures.get(), error.get()));
```

Comment:
Failed to publish %d records.

Reply:
the error should be passed as the cause of the IOException, not included in the error message.